MFVNeutralino/plugins/Vertexer.cc

```diff
@@ -1,891 +1,1778 @@
+
+  std::vector<std::vector<size_t>> sv_lonesharedtrack_which_trkidx;
+  std::vector<size_t> sv0_lonesharedtrack_which_trkidx;
+  std::vector<size_t> sv1_lonesharedtrack_which_trkidx;
+
+  if (shared_jet) {
+
+    int nsharedjets = 0;
+
+    std::vector<std::vector<int>> sv_match_tracktojet_which_temp_jetidx = sv_match_tracktojet_which_jetidx;
+    std::vector<std::vector<int> >sv0_match_sharedjettotrack_which_trkidx;
+    std::vector<std::vector<int> >sv1_match_sharedjettotrack_which_trkidx;
+
+    std::vector<int> sv0_match_tracktojet_which_jetidx = sv_match_tracktojet_which_jetidx[vtx0idx];
+    std::vector<int> sv0_match_tracktojet_which_trkidx = sv_match_tracktojet_which_trkidx[vtx0idx];
+    std::vector<int> sv0_match_sharedjettotrack_which_temp_trkidx;
+
+    std::vector<int> sv1_match_tracktojet_which_jetidx = sv_match_tracktojet_which_jetidx[vtx1idx];
+    std::vector<int> sv1_match_tracktojet_which_trkidx = sv_match_tracktojet_which_trkidx[vtx1idx];
+    std::vector<int> sv1_match_sharedjettotrack_which_temp_trkidx;
+
+    while (std::find_first_of(sv_match_tracktojet_which_temp_jetidx[vtx0idx].begin(), sv_match_tracktojet_which_temp_jetidx[vtx0idx].end(), sv_match_tracktojet_which_temp_jetidx[vtx1idx].begin(), sv_match_tracktojet_which_temp_jetidx[vtx1idx].end()) != sv_match_tracktojet_which_temp_jetidx[vtx0idx].end()) {
+      nsharedjets++;
+      std::vector<int>::iterator it = std::find_first_of(sv_match_tracktojet_which_temp_jetidx[vtx0idx].begin(), sv_match_tracktojet_which_temp_jetidx[vtx0idx].end(), sv_match_tracktojet_which_temp_jetidx[vtx1idx].begin(), sv_match_tracktojet_which_temp_jetidx[vtx1idx].end());
+      int it_idx = std::distance(sv_match_tracktojet_which_temp_jetidx[vtx0idx].begin(), it);
+      int jet_index = sv_match_tracktojet_which_temp_jetidx[vtx0idx].at(it_idx);
+
+      sv_match_tracktojet_which_temp_jetidx[vtx0idx].erase(std::remove(sv_match_tracktojet_which_temp_jetidx[vtx0idx].begin(), sv_match_tracktojet_which_temp_jetidx[vtx0idx].end(), jet_index), sv_match_tracktojet_which_temp_jetidx[vtx0idx].end());
+      sv_match_tracktojet_which_temp_jetidx[vtx1idx].erase(std::remove(sv_match_tracktojet_which_temp_jetidx[vtx1idx].begin(), sv_match_tracktojet_which_temp_jetidx[vtx1idx].end(), jet_index), sv_match_tracktojet_which_temp_jetidx[vtx1idx].end());
+
+      // start collecting shared tracks of sv0 for each shared jet
+      std::multimap<int, size_t> sv0_m;
+      for (size_t k = 0; k < sv0_match_tracktojet_which_jetidx.size(); k++)
+        if (sv0_match_tracktojet_which_jetidx[k] == jet_index) { sv0_m.insert({ sv0_match_tracktojet_which_jetidx[k], k }); }
+
+      for (auto it = sv0_m.begin(); it != sv0_m.end(); )
+      {
+        auto p = sv0_m.equal_range(it->first);
+
+        while (p.first != p.second)
+        {
+
+          sv0_match_sharedjettotrack_which_temp_trkidx.push_back(sv0_match_tracktojet_which_trkidx[p.first++->second]);

```

Comment:
This is pretty opaque to me, especially `p.first++->second`. Can we do this in an equivalent way that is a bit easier to interpret?

Reply:
I got this algorithm for the internet long time ago, and I was wondering if we can treat it as a black box? just like other tools we have.

Reply:
We should try to avoid having any significant parts of our code be black boxes--it's a lot harder to debug when issues arise (and surprises always will arise). It's okay to have a comment that describes what is happening (perhaps including a link to where the algorithm came from), if rewriting it would be too much work. Otherwise if we are trying to investigate some surprise later on, we could e.g. look at this code snippet, think that our "black box" had a typo, and make some change that introduces real issues.

And if there are other parts of the code that you think are too opaque currently / seem like a black box, we can and should make improvements! The vertexer is arguably the most important part of our code, since it is the algorithm that actually creates our displaced vertex objects, so I think it is extremely important for it to be understandable and robust.

Reply:
I add comments and FIXME for this part. It might be able to simplify but we will have to investigate for an improvement.

Reply:
Okay, I have made a function to clarify what it is doing for now.